Add sprint 2 iteration 9

a-tho opened the pull request.

a-tho committed:
Log handlers with decorator

a-tho committed:
Add new JSON-handling handlers and refactor

a-tho committed:
Remove trailing slashes for new endpoints

a-tho committed:
Use 404 instead of 415

a-tho committed:
Add trailing slash for new endpoints

a-tho committed:
Add trailing slash for new endpoint /value/

a-tho committed:
agent: No longer exit when failed to send metrics

a-tho committed:
Add compression

a-tho committed:
Include Content-Encoding header for all resp codes

a-tho committed:
Return Content-Encoding header for 200 responses

a-tho committed:
Add optional compression for / endpoint

a-tho committed:
Return text/html for / endpoint

a-tho committed:
Do not expect request to have extra headers

a-tho committed:
Write metrics to file

a-tho committed:
Check if file is initialized properly

a-tho committed:
Simplify StoreInterval logic

a-tho committed:
Update file permissions

a-tho committed:
Make app exit when server fails to launch

a-tho committed:
Log flags

a-tho committed:
Use int instead of time.Duration for -i flag

rktkov commented on internal/server/server.go:
```diff
@@ -6,28 +6,38 @@ import (
 	"fmt"
 
 	"github.com/go-chi/chi/v5"
+	"github.com/rs/zerolog"
```
Я бы детали реализации логгирования скрыл бы за отдельным пакетом, внутри которого уже бы использовал внешний.

rktkov commented on internal/server/logging.go:
```diff
@@ -0,0 +1,51 @@
+package server
```
Стоит сделать отдельный пакет middleware

rktkov commented on internal/server/compress.go:
```diff
@@ -0,0 +1,93 @@
+package server
```
Стоит сделать отдельный пакет middleware

rktkov commented on internal/monitor.go:
```diff
@@ -17,15 +24,14 @@ type MetricRepo interface {
+
+	// MarshalJSON() ([]byte, error)
+	// UnmarshalJSON(data []byte) error
+	WriteToFile() error
```
Выглядит как лишний метод в интерфейс, обязывает все имплементации пакета писать что-то в файл.

rktkov commented on cmd/server/main.go:
```diff
@@ -13,7 +17,12 @@ import (
 type Config struct {
 	// Flags
-	SrvAddr string `env:"ADDRESS"`
+	SrvAddr         string `env:"ADDRESS"`
```
Стоит сделать отдельный пакет для работы с конфигурацией/настройками.

a-tho committed:
Refactor

- No longer expose WriteToFile()
- use mutex to prevent concurrent access to map

a-tho committed:
Add middleware package

a-tho committed:
Add config package
